datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKey.kt

```diff
@@ -1,117 +1,182 @@
+                if (other.x == null) return false
+                if (!x.contentEquals(other.x)) return false
+            } else if (other.x != null) return false
+            if (y != other.y) return false
+            if (d != null) {
+                if (other.d == null) return false
+                if (!d.contentEquals(other.d)) return false
+            } else if (other.d != null) return false
+
+            return true
+        }
+
+        override fun hashCode(): Int {
+            var result = curve?.hashCode() ?: 0
+            result = 31 * result + (x?.contentHashCode() ?: 0)
+            result = 31 * result + (y?.hashCode() ?: 0)
+            result = 31 * result + (d?.contentHashCode() ?: 0)
+            return result
+        }
+
+        override fun toCryptoPublicKey(): CryptoPublicKey? = TODO()
+
+//        TODO conversion to cryptoPublicKey (needs de-/compression of Y coordinate)
+    }
+
+    // Implements RSA public key parameters
+    data class RsaParams(
+        val n: ByteArray? = null,
+        val e: ByteArray? = null,
+        val d: ByteArray? = null
+    ) : CoseKeyParams() {
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (other == null || this::class != other::class) return false
+
+            other as RsaParams
+
+            if (n != null) {
+                if (other.n == null) return false
+                if (!n.contentEquals(other.n)) return false
+            } else if (other.n != null) return false
+            if (e != null) {
+                if (other.e == null) return false
+                if (!e.contentEquals(other.e)) return false
+            } else if (other.e != null) return false
+            if (d != null) {
+                if (other.d == null) return false
+                if (!d.contentEquals(other.d)) return false
+            } else if (other.d != null) return false
+
+            return true
+        }
+
+        override fun hashCode(): Int {
+            var result = n?.contentHashCode() ?: 0
+            result = 31 * result + (e?.contentHashCode() ?: 0)
+            result = 31 * result + (d?.contentHashCode() ?: 0)
+            return result
+        }
+
+        override fun toCryptoPublicKey(): CryptoPublicKey? {
+            return let {
+                CryptoPublicKey.Rsa(
+                    n = n ?: return null,
+                    e = e?.let { bytes -> Int.decodeFromDer(bytes) } ?: return null
+                )
             }
-            val keyId = MultibaseHelper.calcKeyId(curve.toJwkCurve(), x, y)
-            return CoseKey(
-                type = type,
-                keyId = keyId.encodeToByteArray(),
-                algorithm = CoseAlgorithm.ES256,
-                curve = curve,
-                x = x,
-                y = y
-            )
         }
     }
+}
+
 
+@OptIn(ExperimentalSerializationApi::class)
+@Serializable(with = CoseKeySerializer::class)
+data class CoseKey(
```

Comment:
Just to increase readability of this huge file, I would either put `CoseKey` first or even move all other classes to a separate file

Reply:
separated CoseKey and CoseKeyParams into two separate files